File: app/src/main/java/com/travlog/android/apps/libs/BaseFragment.java
package com.travlog.android.apps.libs;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.CallSuper;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.travlog.android.apps.FragmentViewModel;
import com.travlog.android.apps.libs.qualifiers.RequiresFragmentViewModel;
import com.travlog.android.apps.libs.utils.BundleUtilsKt;
import com.trello.rxlifecycle2.components.support.RxFragment;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;

public class BaseFragment<ViewModelType extends FragmentViewModel> extends RxFragment
        implements FragmentLifeCycleType {

    private static final String VIEW_MODEL_KEY = "viewModel";
    private final CompositeDisposable disposables = new CompositeDisposable();
    private Unbinder unbinder;
    protected ViewModelType viewModel;

    public ViewModelType viewModel() {
        return this.viewModel;
    }

    @CallSuper
    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        Timber.d("onAttach %s", this.toString());
    }

    @CallSuper
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Timber.d("onCreate %s", this.toString());
        assignViewModel(savedInstanceState);
    }

    @CallSuper
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        Timber.d("onCreateView %s", this.toString());
        if (viewModel != null)
            viewModel.onCreateView(inflater, container, savedInstanceState, this);


        return super.onCreateView(inflater, container, savedInstanceState);
    }

    protected View setContentView(@LayoutRes int layoutResID) {
        final View view = LayoutInflater.from(getContext()).inflate(layoutResID, null);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @CallSuper
    @Override
    public void onStart() {
        super.onStart();
        Timber.d("onStart %s", this.toString());
        if (viewModel != null)
            viewModel.onStart();
    }

    @CallSuper
    @Override
    public void onResume() {
        super.onResume();
        Timber.d("onResume %s", this.toString());
        if (viewModel != null)
            viewModel.onResume();
    }

    @CallSuper
    @Override
    public void onPause() {
        super.onPause();
        Timber.d("onPause %s", this.toString());
        if (viewModel != null)
            viewModel.onPause();
    }

    @CallSuper
    @Override
    public void onStop() {
        super.onStop();
        Timber.d("onStop %s", this.toString());
        if (viewModel != null)
            viewModel.onStop();
    }

    @CallSuper
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Timber.d("onDestroyView %s", this.toString());
        if (viewModel != null)
            viewModel.onDestroyView();
        unbinder.unbind();
    }

    @CallSuper
    @Override
    public void onDestroy() {
        super.onDestroy();
        Timber.d("onDestroy %s", this.toString());
        if (viewModel != null)
            viewModel.onDestroy();
    }

    @CallSuper
    @Override
    public void onDetach() {
        Timber.d("onDetach %s", this.toString());
        super.onDetach();

        if (getActivity().isFinishing()) {
            if (viewModel != null) {
                viewModel.onDetach();
            }
        }
    }

    private void assignViewModel(final @Nullable Bundle viewModelEnvelope) {
        if (viewModel == null) {
            final RequiresFragmentViewModel annotation = getClass().getAnnotation(RequiresFragmentViewModel.class);
            final Class<ViewModelType> ViewModelTypeClass = annotation == null ? null : (Class<ViewModelType>) annotation.value();
            if (ViewModelTypeClass != null) {
                viewModel = FragmentViewModelManager.getInstance().fetch(getContext(),
                        ViewModelTypeClass,
                        BundleUtilsKt.maybeGetBundle(viewModelEnvelope, VIEW_MODEL_KEY));
            }
        }
    }
}

File: app/src/main/java/com/travlog/android/apps/libs/BaseFragment.kt
package com.travlog.android.apps.libs

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import com.travlog.android.apps.FragmentViewModel
import com.trello.rxlifecycle3.components.support.RxFragment
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber

open class BaseFragment<ViewModelType : FragmentViewModel<*>> : RxFragment(), FragmentLifeCycleType {
    private val disposables = CompositeDisposable()
    private val VIEW_MODEL_KEY = "viewModel"
    protected var viewModel: ViewModelType? = null

    fun viewModel(): ViewModelType? {
        return this.viewModel
    }

    @CallSuper
    override fun onAttach(context: Context) {
        super.onAttach(context)
        Timber.d("onAttach %s", this.toString())
    }

    @CallSuper
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Timber.d("onCreate %s", this.toString())
    }

    @CallSuper
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        Timber.d("onCreateView %s", this.toString())

        viewModel?.onCreateView(inflater, container, savedInstanceState, this)

        return super.onCreateView(inflater, container, savedInstanceState)
    }

    @CallSuper
    override fun onStart() {
        super.onStart()
        Timber.d("onStart %s", this.toString())
        viewModel?.onStart()
    }

    @CallSuper
    override fun onResume() {
        super.onResume()
        Timber.d("onResume %s", this.toString())
        viewModel?.onResume()
    }

    @CallSuper
    override fun onPause() {
        super.onPause()
        Timber.d("onPause %s", this.toString())
        viewModel?.onPause()
    }

    @CallSuper
    override fun onStop() {
        super.onStop()
        Timber.d("onStop %s", this.toString())
        viewModel?.onStop()
    }

    @CallSuper
    override fun onDestroyView() {
        super.onDestroyView()
        Timber.d("onDestroyView %s", this.toString())
        viewModel?.onDestroyView()
    }

    @CallSuper
    override fun onDestroy() {
        super.onDestroy()
        Timber.d("onDestroy %s", this.toString())
        viewModel?.onDestroy()
    }

    @CallSuper
    override fun onDetach() {
        Timber.d("onDetach %s", this.toString())
        super.onDetach()

        activity?.apply {
            if (isFinishing) {
                viewModel?.onDetach()
            }
        }
    }
}
